libsnark/gadgetlib1/gadgets/pairing/bw6_761_bls12_377/bw6_761_pairing_params.hpp

```diff
@@ -0,0 +1,88 @@
+
+#include "libsnark/gadgetlib1/gadgets/fields/fp12_2over3over2_gadgets.hpp"
+#include "libsnark/gadgetlib1/gadgets/fields/fp2_gadgets.hpp"
+#include "libsnark/gadgetlib1/gadgets/pairing/bw6_761_bls12_377/bls12_377_membership_check_gadgets.hpp"
+#include "libsnark/gadgetlib1/gadgets/pairing/bw6_761_bls12_377/bls12_377_pairing.hpp"
+#include "libsnark/gadgetlib1/gadgets/pairing/pairing_params.hpp"
+
+#include <libff/algebra/curves/bls12_377/bls12_377_pp.hpp>
+#include <libff/algebra/curves/bw6_761/bw6_761_pp.hpp>
+
+namespace libsnark
+{
+
+template<typename ppT> class bls12_377_G1_precomputation;
+template<typename ppT> class bls12_377_G1_precompute_gadget;
+template<typename ppT> class bls12_377_G2_precomputation;
+template<typename ppT> class bls12_377_G2_precompute_gadget;
+template<typename ppT>
+class bls12_377_e_times_e_times_e_over_e_miller_loop_gadget;
+template<typename ppT> class bls12_377_final_exp_gadget;
+
+// Parameters for creating BW6-761 proofs that include statements about
+// BLS12_377 pairings.
+template<> class pairing_selector<libff::bw6_761_pp>
+{
+public:
+    static_assert(
+        std::is_same<
+            libff::Fr<libff::bw6_761_pp>,
+            libff::Fq<libff::bls12_377_pp>>::value,
+        "Field types do not match");
+
+    typedef libff::Fr<libff::bw6_761_pp> FieldT;
+    typedef libff::Fqe<libff::bls12_377_pp> FqeT;
+    typedef libff::Fqk<libff::bls12_377_pp> FqkT;
+
+    typedef Fp2_variable<FqeT> Fqe_variable_type;
+    typedef Fp2_mul_gadget<FqeT> Fqe_mul_gadget_type;
+    typedef Fp2_mul_by_lc_gadget<FqeT> Fqe_mul_by_lc_gadget_type;
+    typedef Fp2_sqr_gadget<FqeT> Fqe_sqr_gadget_type;
+
+    typedef Fp12_2over3over2_variable<FqkT> Fqk_variable_type;
+    typedef Fp12_2over3over2_mul_gadget<FqkT> Fqk_mul_gadget_type;
+    typedef Fp12_2over3over2_square_gadget<FqkT> Fqk_sqr_gadget_type;
+
+    typedef libff::bls12_377_pp other_curve_type;
+
+    typedef bls12_377_G1_membership_check_gadget<libff::bw6_761_pp>
+        G1_checker_type;
+    typedef bls12_377_G2_membership_check_gadget<libff::bw6_761_pp>
+        G2_checker_type;
+
+    typedef bls12_377_G1_precomputation<libff::bw6_761_pp>
+        G1_precomputation_type;
+    typedef bls12_377_G1_precompute_gadget<libff::bw6_761_pp>
+        precompute_G1_gadget_type;
+
+    typedef bls12_377_G2_precomputation<libff::bw6_761_pp>
+        G2_precomputation_type;
+    typedef bls12_377_G2_precompute_gadget<libff::bw6_761_pp>
+        precompute_G2_gadget_type;
+
+    // Not implemented:
+    // typedef bls12_377_e_over_e_miller_loop_gadget
+    //     bls12_377_over_e_miller_loop_gadget_type;
+    // typedef bls12_377_e_times_e_over_e_miller_loop_gadget
+    //     e_times_e_over_e_miller_loop_gadget_type;
+
+    typedef bls12_377_e_times_e_times_e_over_e_miller_loop_gadget<
+        libff::bw6_761_pp>
+        e_times_e_times_e_over_e_miller_loop_gadget_type;
+
+    typedef bls12_377_final_exp_gadget<libff::bw6_761_pp> final_exp_gadget_type;
+};
+
```

Comment:
> OK. If we can't think of a good reason I'd prefer to remove it (TBH, if it's required as a cross-curve parameter, I would expect it to be in libff anyway). Also removed another special_mul gadget for one of the fields/

> AFAICT, these were only here so that the mnt code itself could select, so I've added a separate parameterized class mnt_pairing_params for internal use in the mnt code.

Ok, let's double check this. I'm not fundamentally against the interface change, but we should ideally - if not causing issues with the PR - do this as part of another PR, and this PR should simply be about moving the code over. Here we mix: code ported from Zecale + interface changes. That would be easier to stick to the interface proposed, move the code and then think about changing the interface. Let's keep it that way for this one, but try to avoid this for further PRs.

Reply:
Right.  It would certainly be preferable to proceed in smaller steps.  The complication here is largely because the code being moved in *required* the interface changes, but as in https://github.com/clearmatics/libsnark/pull/29#issuecomment-948497285, it may be possible to extract the interface changes *first* in one PR, and then move the code in a *second*.  Happy to try this if reviewing this PR becomes too painful.